Emby.Server.Implementations/Chapters/ChapterManager.cs

```diff
@@ -0,0 +1,316 @@
+/// </summary>
+public class ChapterManager : IChapterManager
+{
+    private readonly IFileSystem _fileSystem;
+    private readonly ILogger<ChapterManager> _logger;
+    private readonly IMediaEncoder _encoder;
+    private readonly IChapterRepository _chapterRepository;
+    private readonly ILibraryManager _libraryManager;
+    private readonly IPathManager _pathManager;
+
+    /// <summary>
+    /// The first chapter ticks.
+    /// </summary>
+    private static readonly long _firstChapterTicks = TimeSpan.FromSeconds(15).Ticks;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="ChapterManager"/> class.
+    /// </summary>
+    /// <param name="logger">The <see cref="ILogger{ChapterManager}"/>.</param>
+    /// <param name="fileSystem">The <see cref="IFileSystem"/>.</param>
+    /// <param name="encoder">The <see cref="IMediaEncoder"/>.</param>
+    /// <param name="chapterRepository">The <see cref="IChapterRepository"/>.</param>
+    /// <param name="libraryManager">The <see cref="ILibraryManager"/>.</param>
+    /// <param name="pathManager">The <see cref="IPathManager"/>.</param>
+    public ChapterManager(
+        ILogger<ChapterManager> logger,
+        IFileSystem fileSystem,
+        IMediaEncoder encoder,
+        IChapterRepository chapterRepository,
+        ILibraryManager libraryManager,
+        IPathManager pathManager)
+    {
+        _logger = logger;
+        _fileSystem = fileSystem;
+        _encoder = encoder;
+        _chapterRepository = chapterRepository;
+        _libraryManager = libraryManager;
+        _pathManager = pathManager;
+    }
+
+    /// <summary>
+    /// Determines whether [is eligible for chapter image extraction] [the specified video].
+    /// </summary>
+    /// <param name="video">The video.</param>
+    /// <param name="libraryOptions">The library options for the video.</param>
+    /// <returns><c>true</c> if [is eligible for chapter image extraction] [the specified video]; otherwise, <c>false</c>.</returns>
+    private bool IsEligibleForChapterImageExtraction(Video video, LibraryOptions libraryOptions)
+    {
+        if (video.IsPlaceHolder)
+        {
+            return false;
+        }
+
+        if (libraryOptions is null || !libraryOptions.EnableChapterImageExtraction)
+        {
+            return false;
+        }
+
+        if (video.IsShortcut)
+        {
+            return false;
+        }
+
+        if (!video.IsCompleteMedia)
+        {
+            return false;
+        }
+
+        // Can't extract images if there are no video streams
+        return video.DefaultVideoStreamIndex.HasValue;
+    }
+
+    private long GetAverageDurationBetweenChapters(IReadOnlyList<ChapterInfo> chapters)
+    {
+        if (chapters.Count < 2)
+        {
+            return 0;
+        }
+
+        long sum = 0;
+        for (int i = 1; i < chapters.Count; i++)
+        {
+            sum += chapters[i].StartPositionTicks - chapters[i - 1].StartPositionTicks;
```

Comment:
shouldent that be instead if you are looking for the duration __between__ chapters?
```suggestion
            sum += chapters[i - 1].EndPositionTicks - chapters[i].StartPositionTicks;
```

Reply:
`StartPositionTicks ` is the marker of the chapter, they don't have a duration.

Reply:
right i thought about MediaSegments.
Wupsi